calculator.go

```diff
@@ -1,13 +1,99 @@
 // Package calculator provides a library for simple calculations in Go.
 package calculator
 
-// Add takes two numbers and returns the result of adding them together.
-func Add(a, b float64) float64 {
-return a + b
+import (
+	"errors"
+	"fmt"
+	"math"
+)
+
+// Add takes two or more numbers and returns the result of adding them together.
+func Add(a, b float64, nums ...float64) float64 {
+	var sum float64 = a + b
+	for _, num := range nums {
+		sum += num
+	}
+	return sum
 }
 
-// Subtract takes two numbers and returns the result of subtracting the second
+// Subtract takes two or more numbers and returns the result of subtracting the second
 // from the first.
-func Subtract(a, b float64) float64 {
-	return b - a
+func Subtract(a, b float64, nums ...float64) float64 {
+	var total float64 = a - b
+	for _, num := range nums {
+		total -= num
+	}
+	return total
+}
+
+// Multiply takes two or more numbers and returns the result of multiplying them together
+func Multiply(a, b float64, nums ...float64) float64 {
+	var total float64 = a * b
+	for _, num := range nums {
+		total *= num
+	}
+	return total
+}
+
+// Divide takes two or more numbers and returns the result of dividing a by b
+func Divide(a, b float64, nums ...float64) (float64, error) {
+	if b == 0 {
+		return 0, errors.New("divide by zero")
+	}
+	// if the first number is zero then the result is zero.
```

Comment:
It is just another example of the early optimization.  I put the optimization in when I added the variadic parameter nums.  If someone passed 1000 numbers to this function then it might be worthwhile to not divide 0 by the following 999 numbers?

Reply:
Yes, absolutely! But it's really useful for us to be able to separate what's essential from what's an optimisation or enhancement, isn't it? So the process is to ruthlessly cut away everything that's not essential, get that complete and working and stabilise the API, and _then_ we can think about what's next—without spending any effort on things which might still change or disappear altogether.